### ISSUE TITLE
New lint: "use Option::take"

### ISSUE THREAD

**Issue by glittershark** [rust-lang/rust-clippy#6711]
### What it does

Suggest replacing `std::mem::replace(&mut x, None)` `x.take()`

### Categories (optional)

- Kind: `clippy::style`

Advantages: A simpler way of accomplishing the same thing

### Drawbacks

None that I can tell

### Example

```rust
let y = std::mem::replace(&x, None);
```

Could be written as:

```rust
let y = x.take()
```

**Comment by giraffate** [comment]
Do you mean [mem_replace_option_with_none](https://rust-lang.github.io/rust-clippy/master/index.html#mem_replace_option_with_none)? It seems to be same as this, so I'm closing this.

**Comment by glittershark** [comment]
Huh. I may have a false negative case for that then. I'll see if I can repro

**Comment by JayKickliter** [comment]
@glittershark were you able to find a repro?

**Comment by kalabukdima** [comment]
I can reproduce it like this: `std::mem::replace(&mut *x, None)`, which doesn't make much sense if `x` is just `&mut Option` (though clippy doesn't warn about it either), but can be used on types with overloaded `DerefMut`. For example:
```rust
pub fn foo(x: Mutex<Option<u32>>) -> u32 {
  std::mem::replace(&mut *x.lock().unwrap(), None).unwrap()
}
```
Clippy doesn't warn about anything here. Should I report a new issue?
